library/Notifications/Widget/Calendar/BaseGrid.php

```diff
@@ -289,16 +340,92 @@ protected function assembleGridOverlay(BaseHtmlElement $overlay): void
                         'data-col-end' => $gridArea[3]
                     ])
                 );
-                $this->assembleEntry($entryHtml, $entry, $continuation);
+
+                $this->assembleEntry($entryHtml, $entry, $startText, $endText, $continuation);
                 $overlay->addHtml($entryHtml);
 
                 $continuation = true;
             }
         }
     }
 
-    protected function assembleEntry(BaseHtmlElement $html, Entry $entry, bool $isContinuation): void
-    {
+    /**
+     * Fetch properties for entryHtml
+     *
+     * @param string $entryColor
+     * @param string|null $direction
+     * @param string|null $opposite
+     *
+     * @return array
+     */
+    protected function getEntryHtmlProperties(
```

Comment:
Please define as much as possible in CSS already. The only variables are the contact's color values. Since you need them now for more than just a background and border color, just set some CSS variables and reference them instead of generating the entire rules here. I suppose once you did this, this method is obsolete.